tests/test_ufrm.py

```diff
@@ -151,6 +154,84 @@ def test_ufrm_value_error_on_bad_soil(self):
                 'Check that the Soil Group raster does not contain')
             self.assertTrue(expected_message in actual_message)
 
+    def test_ufrm_string_damage_to_infrastructure(self):
+        """UFRM: handle str(int) structure indices.
+
+        This came up on the forums, where a user had provided a string column
+        type that contained integer data.  OGR returned these ints as strings,
+        leading to a ``KeyError``.  See
+        https://github.com/natcap/invest/issues/590.
+        """
+        from natcap.invest import urban_flood_risk_mitigation
+
+        srs = osr.SpatialReference()
+        srs.ImportFromEPSG(3157)
+        projection_wkt = srs.ExportToWkt()
+        origin = (443723.127327877911739, 4956546.905980412848294)
+        pos_x = origin[0]
+        pos_y = origin[1]
+
+        aoi_geometry = [
+            shapely.geometry.box(pos_x, pos_y, pos_x + 200, pos_y + 200),
+        ]
+
+        def _infra_geom(xoff, yoff):
+            """Create sample infrastructure geometry at a position offset.
+
+            The geometry will be centered on (x+xoff, y+yoff).
+
+            Parameters:
+                xoff (number): The x offset, referenced against ``pos_x`` from
+                    the outer scope.
+                yoff (number): The y offset, referenced against ``pos_y`` from
+                    the outer scope.
+
+            Returns:
+                A ``shapely.Geometry`` of a point buffered by ``20`` centered
+                on the provided (x+xoff, y+yoff) point.
+            """
+            return shapely.geometry.Point(
+                pos_x + xoff, pos_y + yoff).buffer(20)
+
+        infra_geometries = [
+            _infra_geom(x_offset, 100)
+            for x_offset in range(0, 200, 40)]
+
+        infra_fields = {'Type': ogr.OFTString}  # THIS IS THE THING TESTED
+        infra_attrs = [
+            {'Type': str(index)} for index in range(len(infra_geometries))]
+
+        infrastructure_path = os.path.join(
+            self.workspace_dir, 'infra_vector.shp')
+        pygeoprocessing.shapely_geometry_to_vector(
+            infra_geometries, infrastructure_path, projection_wkt,
+            'ESRI Shapefile', fields=infra_fields, attribute_list=infra_attrs,
+            ogr_geom_type=ogr.wkbPolygon)

```

Comment:
I'm probably just confused, but we're creating buffered infrastructure points along a line (because `yoff` stays at `100`). Then we take those geoms and make a vector type Polygon. Wouldn't that just be a line and the buffered points be unnecessary? Sorry if I'm completely missing something with how the infrastructure vector should be set up.

Reply:
Thanks for taking a look at this!

You're very right that they're along a line, but the circles are at 40-unit intervals along that line (see line 198) so the end result looks like this:

<img width="697" alt="Screen Shot 2021-07-21 at 12 55 46 PM" src="https://user-images.githubusercontent.com/4583897/126551937-8df07051-4ac0-469b-8b56-9f15ab385d43.png">

Reply:
Ah right, sorry, my brain failed to make the buffered points to polygon connection.